test/zhuyin_test.exs

```diff
@@ -0,0 +1,112 @@
+defmodule ZhuyinTest do
+  use ExUnit.Case
+  import Zhuyin
+  import Pinyin
+  doctest Zhuyin
+
+  describe "Decode Zhuyin" do
+    test "Full syllables" do
+      tests = [
+        {"ㄐㄩ˙", ~p/jv/},
+        {"ㄌㄩˇ", ~p/lv3/},
+        {"ㄓㄠˊ", ~p/zhao2/},
+        {"ㄓˋ", ~p/zhi4/},
+        {"ㄌㄥ", ~p/leng1/},
+        {"ㄕㄨㄟˇ", ~p/shui3/},
+        {"ㄌㄧㄡˊ", ~p/liu2/},
+        {"ㄧˊ", ~p/yi2/},
+        {"ㄇㄧㄣˊ", ~p/min2/},
+        {"ㄨㄥˊ", ~p/weng2/},
+        {"ㄨˊ", ~p/wu2/},
+        {"ㄩ", ~p/yu1/},
+        {"ㄩㄝˇ", ~p/yue3/},
+        {"ㄩㄝˋ", ~p/yue4/},
+        {"ㄩㄢˊ", ~p/yuan2/},
+        {"ㄌㄩㄢˋ", ~p/lvan4/},
+        {"ㄌㄢˇ", ~p/lan3/},
+        {"ㄦˊ", ~p/er2/}
+        # Pinyin parser can't parse r right now.
+        # ㄦ/兒 in first tone is just "r" in pinyin
+        # {"ㄦ˙", ~p/r0/}
+      ]
+
+      Enum.each(tests, fn {z, p} -> assert Zhuyin.read!(z) |> Zhuyin.to_pinyin() == p end)
+    end
+
+    test "Incomplete syllables" do
+      tests = [
+        "ㄝ",
+        "ㄐˇ"
+      ]
+
+      Enum.each(tests, fn z -> assert Zhuyin.read(z) == {:error, z} end)
+    end
+
+    test "Standalone initials" do
+      # First a regular syllable
+      {:ok, [%Zhuyin{initial: "ㄋ", final: "ㄧ", tone: 3}], "", %{}, {1, 0}, 8} =
+        Zhuyin.Parsers.syllable("ㄋㄧˇ")
+
+      # Must not allow standalone initials be combined with a final
+      assert {:error, "expected end of string", "ㄧˇ", %{}, {1, 0}, 3} ==
+               Zhuyin.Parsers.syllable("ㄕㄧˇ")
+
+      # That can only be parsed if we allow multiple words
+      assert {:ok, [~z/ㄕ/, ~z/ㄧˇ/] |> List.flatten(), "", %{}, {1, 0}, 8} ==
+               Zhuyin.Parsers.zhuyin_only("ㄕㄧˇ")
```

Comment:
I'm changing the tests to use `Zhuyin.read` and others instead of the parsers, as I consider these to be an implementation detail.